src/linear_algebra/quadratic_form.lean

```diff
@@ -0,0 +1,290 @@
+/-
+Copyright (c) 2020 Anne Baanen. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Author: Anne Baanen
+
+Quadratic forms over modules.
+-/
+
+import linear_algebra.bilinear_form
+import linear_algebra.determinant
+import linear_algebra.special_linear_group
+
+/-!
+# Quadratic forms
+
+This file defines quadratic forms over a `R`-module `M`.
+A quadratic form is a map `Q : M → R` such that
+  (`map_smul`) `Q (a • x) = a * a * Q x`
+  (`polar_...`) The map `polar Q := λ x y, Q (x + y) - Q x - Q y` is bilinear.
+They come with a scalar multiplication, `(a • Q) x = Q (a • x) = a * a * Q x`,
+and composition with linear maps `f`, `Q.comp f x = Q (f x)`.
+
+## Main definitions
+
+ * `quadratic_form.associated`: associated bilinear form
+ * `quadratic_form.pos_def`: positive definite quadratic forms
+ * `quadratic_form.discr`: discriminant of a quadratic form
+
+## Main statements
+
+ * `quadratic_form.associated_left_inverse`,
+ * `quadratic_form.associated_right_inverse`: in a commutative ring where 2 has
+  an inverse, there is a correspondence between quadratic forms and symmetric
+  bilinear forms
+
+## References
+
+ * https://en.wikipedia.org/wiki/Quadratic_form
+ * https://en.wikipedia.org/wiki/Discriminant#Quadratic_forms
+
+## Tags
+
+quadratic form, homogeneous polynomial, quadratic polynomial
+-/
+
+universes u v w
+variables {R : Type u} {M : Type v} [add_comm_group M] [ring R]
+variables {R₁ : Type u} [comm_ring R₁] [module R₁ M]
+
+/-- Up to a factor 2, `Q.polar` is the associated bilinear form for a quadratic form `Q`. -/
+def quadratic_form.polar (f : M → R) (x y : M) :=
+f (x + y) - f x - f y
+
+variables [module R M]
+
+open quadratic_form
+/-- A quadratic form over a module. -/
+structure quadratic_form (R : Type u) (M : Type v) [ring R] [add_comm_group M] [module R M] :=
+(map : M → R)
+(map_smul : ∀ (a : R) (x : M), map (a • x) = a * a * map x)
+(polar_add_left : ∀ (x x' y : M), polar map (x + x') y = polar map x y + polar map x' y)
+(polar_smul_left : ∀ (a : R) (x y : M), polar map (a • x) y = a • polar map x y)
+(polar_add_right : ∀ (x y y' : M), polar map x (y + y') = polar map x y + polar map x y')
+(polar_smul_right : ∀ (a : R) (x y : M), polar map x (a • y) = a • polar map x y)
+
+namespace quadratic_form
+
+variables {Q : quadratic_form R M}
+
+instance : has_coe_to_fun (quadratic_form R M) :=
+⟨_, λ B, B.map⟩
+
+/-- The `simp` normal form for a quadratic form is `coe_fn`, not `map`. -/
+@[simp] lemma map_eq_apply : Q.map = ⇑ Q := rfl
+
+lemma apply_smul (a : R) (x : M) : Q (a • x) = a * a * Q x := Q.map_smul a x
```

Comment:
I think this should be called `map_smul`, to match what happens elsewhere. You could put a prime on the axioms in the structure to avoid a name clash.

Reply:
I've renamed `quadratic_form.map` to `quadratic_form.to_fun` as you suggested. So the changes are `apply_smul -> map_smul` and `map_smul -> to_fun_smul`